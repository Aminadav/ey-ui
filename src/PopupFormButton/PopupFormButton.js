import React, { useState } from 'react';
import Popover from '@material-ui/core/Popover';
import PropTypes from 'prop-types';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';
import useStyles from './styles';

const PopupFormButton = props => {
  const {
    Form,
    Icon,
    FormProps,
    buttonText,
    ButtonProps: {
      variant: buttonVariant = 'outlined',
      color: buttonColor = 'default',
      ...ButtonProps
    },
  } = props;
  const classes = useStyles();
  const [isOpen, setIsOpen] = useState(false);
  const [anchor, setAnchor] = useState(null);

  const anchorOrigin = {
    vertical: 'bottom',
    horizontal: 'center',
  };

  const transformOrigin = {
    vertical: 'top',
    horizontal: 'center',
  };

  const handleClick = event => {
    setAnchor(event.target);
    setIsOpen(!isOpen);
  };

  const handleClose = () => setIsOpen(false);

  return (
    <React.Fragment>
      <Popover
        anchorEl={anchor}
        open={isOpen}
        onClose={() => setIsOpen(false)}
        anchorOrigin={anchorOrigin}
        transformOrigin={transformOrigin}
        data-testid="popover"
      >
        <Form onCancel={handleClose} {...FormProps} />
      </Popover>

      {!buttonText && Icon ? (
        <IconButton data-testid="icon-button" onClick={handleClick} {...ButtonProps}>
          <Icon />
        </IconButton>
      ) : (
        <Button
          data-testid="popup-button"
          variant={buttonVariant}
          color={buttonColor}
          onClick={handleClick}
          {...ButtonProps}
        >
          {buttonText}
          {Icon && <Icon data-testid="icon-in-button" className={classes.rightIcon} />}
        </Button>
      )}
    </React.Fragment>
  );
};

PopupFormButton.propTypes = {
  Form: PropTypes.elementType.isRequired,
  ButtonProps: PropTypes.shape({
    variant: PropTypes.string,
    color: PropTypes.string,
  }),
  buttonText: PropTypes.string,
  Icon: PropTypes.elementType,
  /**
   * Any props you want to pass into the form component.
   */
  FormProps: PropTypes.shape({}),
};

PopupFormButton.defaultProps = {
  buttonText: null,
  ButtonProps: {},
  Icon: null,
  FormProps: {},
};

export default PopupFormButton;
