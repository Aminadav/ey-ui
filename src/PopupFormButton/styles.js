import { makeStyles } from '@material-ui/styles';

const useStyles = makeStyles(theme => ({
  rightIcon: {
    marginLeft: theme.spacing(1),
  },
}));

export default useStyles;
