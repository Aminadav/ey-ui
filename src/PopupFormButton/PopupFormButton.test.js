import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import AddCircle from '@material-ui/icons/AddCircle';
import PopupFormButton from './PopupFormButton';

const Form = () => <form />;

describe('PopupFormButton', () => {
  it('matches the snapshot', () => {
    const { container } = render(<PopupFormButton Form={Form} buttonText="Submit" />);
    expect(container).toMatchSnapshot();
  });

  it('should display a button with proper text conteact', () => {
    const { getByTestId } = render(<PopupFormButton Form={Form} buttonText="Submit" />);

    expect(getByTestId('popup-button')).toHaveTextContent('Submit');
  });

  it('renders an icon with an icon component passed in, and no button text', () => {
    const { getByTestId, queryByTestId } = render(<PopupFormButton Icon={AddCircle} Form={Form} />);
    expect(getByTestId('icon-button')).toBeInTheDocument();
    expect(queryByTestId('popup-button')).toBeNull();
  });

  it('renders button text with button text passed in, and without an icon', () => {
    const { getByTestId } = render(<PopupFormButton buttonText="Test Title" Form={Form} />);

    expect(getByTestId('popup-button')).toHaveTextContent('Test Title');
  });

  it('renders with no icon', () => {
    const { queryByTestId } = render(<PopupFormButton buttonText="Test Title" Form={Form} />);

    expect(queryByTestId('icon-button')).toBeNull();
  });

  it('renders button text & an icon with both passed in', () => {
    const { getByTestId } = render(
      <PopupFormButton buttonText="Test Title" Form={Form} Icon={AddCircle} />,
    );
    expect(getByTestId('icon-in-button')).toBeInTheDocument();
    expect(getByTestId('popup-button')).toHaveTextContent('Test Title');
  });

  it('should display the popover when clicking the icon', () => {
    const { getByTestId, queryByTestId } = render(<PopupFormButton Icon={AddCircle} Form={Form} />);
    const icon = getByTestId('icon-button');
    expect(queryByTestId('popover')).toBeNull();
    fireEvent.click(icon);
    expect(getByTestId('popover')).toBeInTheDocument();
  });

  it('should display the popover when clicking the button', () => {
    const { getByTestId, queryByTestId } = render(
      <PopupFormButton buttonText="Create" Form={Form} />,
    );
    const button = getByTestId('popup-button');
    expect(queryByTestId('popover')).toBeNull();
    fireEvent.click(button);
    expect(getByTestId('popover')).toBeInTheDocument();
  });
});
