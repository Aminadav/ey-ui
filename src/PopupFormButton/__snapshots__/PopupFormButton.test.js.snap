// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupFormButton matches the snapshot 1`] = `
<div>
  <button
    class="MuiButtonBase-root-31 MuiButton-root-5 MuiButton-outlined-13"
    data-testid="popup-button"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-6"
    >
      Submit
    </span>
    <span
      class="MuiTouchRipple-root-34"
    />
  </button>
</div>
`;
